librz/include/rz_type.h

```diff
@@ -0,0 +1,238 @@
+
+typedef struct rz_type_enum_case_t {
+	char *name;
+	int val;
+} RzTypeEnumCase;
+
+typedef struct rz_type_struct_member_t {
+	char *name;
+	char *type;
+	size_t offset; // in bytes
+	size_t size; // in bits?
+} RzTypeStructMember;
+
+typedef struct rz_type_union_member_t {
+	char *name;
+	char *type;
+	size_t offset; // in bytes
+	size_t size; // in bits?
+} RzTypeUnionMember;
+
+typedef enum {
+	RZ_BASE_TYPE_KIND_STRUCT,
+	RZ_BASE_TYPE_KIND_UNION,
+	RZ_BASE_TYPE_KIND_ENUM,
+	RZ_BASE_TYPE_KIND_TYPEDEF, // probably temporary addition, dev purposes
+	RZ_BASE_TYPE_KIND_ATOMIC, // For real atomic base types
+} RzBaseTypeKind;
+
+typedef struct rz_base_type_struct_t {
+	RzVector /*<RzTypeStructMember>*/ members;
+} RzBaseTypeStruct;
+
+typedef struct rz_base_type_union_t {
+	RzVector /*<RzTypeUnionMember>*/ members;
+} RzBaseTypeUnion;
+
+typedef struct rz_base_type_enum_t {
+	RzVector /*<RzTypeEnumCase*/ cases; // list of all the enum casessssss
+} RzBaseTypeEnum;
+
+typedef struct rz_base_type_t {
+	char *name;
+	char *type; // Used by typedef, atomic type, enum
+	ut64 size; // size of the whole type in bits
+	RzBaseTypeKind kind;
+	union {
+		RzBaseTypeStruct struct_data;
+		RzBaseTypeEnum enum_data;
+		RzBaseTypeUnion union_data;
+	};
+} RzBaseType;
+
+typedef struct rz_type_enum {
+	const char *name;
+	const char *val;
+} RzTypeEnum;
+
+#ifdef RZ_API
+
+RZ_API RzTypeDB *rz_type_db_new();
+RZ_API void rz_type_db_free(RzTypeDB *typedb);
+RZ_API void rz_type_db_load_sdb(RzTypeDB *typedb, const char *dpath);
+RZ_API void rz_type_db_purge(RzTypeDB *typedb);
+RZ_API void rz_type_db_set_bits(RzTypeDB *typedb, int bits);
+RZ_API void rz_type_db_set_os(RzTypeDB *typedb, const char *os);
+RZ_API void rz_type_db_set_cpu(RzTypeDB *typedb, const char *cpu);
+RZ_API void rz_type_db_set_endian(RzTypeDB *typedb, bool big_endian);
+RZ_API char *rz_type_db_kuery(RzTypeDB *typedb, const char *query);
+
+RZ_API const char *rz_type_db_get(RzTypeDB *typedb, const char *name);
+RZ_API bool rz_type_db_set(RzTypeDB *typedb, ut64 at, const char *field, ut64 val);
+RZ_API bool rz_type_db_del(RzTypeDB *typedb, RZ_NONNULL const char *name);
+
+RZ_API void rz_type_db_init(RzTypeDB *typedb, const char *dir_prefix, const char *arch, int bits, const char *os);
+
+// Base types
+
+RZ_API void rz_type_base_type(const RzTypeDB *typedb, const RzBaseType *type);
+RZ_API void rz_type_base_type_free(RzBaseType *type);
+RZ_API RzBaseType *rz_type_base_type_new(RzBaseTypeKind kind);
+
+RZ_API void rz_type_base_enum_case_free(void *e, void *user);
+RZ_API void rz_type_base_struct_member_free(void *e, void *user);
+RZ_API void rz_type_base_union_member_free(void *e, void *user);
+
+RZ_API RzBaseType *rz_type_db_get_base_type(RzTypeDB *typedb, const char *name);
+RZ_API void rz_type_db_save_base_type(const RzTypeDB *typedb, const RzBaseType *type);
+
+/* ctype */
+// Parses strings like "const char * [0x42] const * [23]" to RzTypeCTypeType
+
+typedef struct rz_type_ctype_t RzTypeCType;
```

Comment:
this "CType" concept IMO requires *a lot* of explanation... What is the difference between type and ctype? Why is there any difference?

Reply:
It was moved from `librz/util/ctype.c` and will be different after Tree-Sitter refactoring. Might be different. Type is more abstract than CType which can be more abstract than BaseType: https://github.com/rizinorg/rizin/issues/371

Reply:
@ret2libc see https://hackmd.io/@thestr4ng3r/H1gMW5-K8#Solution-Base-Types--C-Types